frame/contracts/src/migration/v12.rs

```diff
@@ -0,0 +1,315 @@
+	#[codec(compact)]
+	refcount: u64,
+	determinism: Determinism,
+}
+
+#[storage_alias]
+pub type CodeInfoOf<T: Config> = StorageMap<Pallet<T>, Twox64Concat, CodeHash<T>, CodeInfo<T>>;
+
+#[storage_alias]
+pub type PristineCode<T: Config> = StorageMap<Pallet<T>, Identity, CodeHash<T>, Vec<u8>>;
+
+#[cfg(feature = "runtime-benchmarks")]
+pub fn store_old_dummy_code<T: Config>(len: usize, account: T::AccountId) {
+	use sp_runtime::traits::Hash;
+
+	let code = vec![42u8; len];
+	let hash = T::Hashing::hash(&code);
+	PristineCode::<T>::insert(hash, code.clone());
+
+	let module = old::PrefabWasmModule {
+		instruction_weights_version: Default::default(),
+		initial: Default::default(),
+		maximum: Default::default(),
+		code,
+		determinism: Determinism::Enforced,
+	};
+	old::CodeStorage::<T>::insert(hash, module);
+
+	let info = old::OwnerInfo { owner: account, deposit: u32::MAX.into(), refcount: u64::MAX };
+	old::OwnerInfoOf::<T>::insert(hash, info);
+}
+
+#[derive(Encode, Decode, MaxEncodedLen, DefaultNoBound)]
+pub struct Migration<T: Config> {
+	last_code_hash: Option<CodeHash<T>>,
+}
+
+impl<T: Config> MigrationStep for Migration<T> {
+	const VERSION: u16 = 12;
+
+	fn max_step_weight() -> Weight {
+		T::WeightInfo::v12_migration_step(T::MaxCodeLen::get())
+	}
+
+	fn step(&mut self) -> (IsFinished, Weight) {
+		let mut iter = if let Some(last_key) = self.last_code_hash.take() {
+			old::OwnerInfoOf::<T>::iter_from(old::OwnerInfoOf::<T>::hashed_key_for(last_key))
+		} else {
+			old::OwnerInfoOf::<T>::iter()
+		};
+		if let Some((hash, old_info)) = iter.next() {
+			log::debug!(target: LOG_TARGET, "Migrating OwnerInfo for code_hash {:?}", hash);
+
+			let module = old::CodeStorage::<T>::take(hash)
+				.expect(format!("No PrefabWasmModule found for code_hash: {:?}", hash).as_str());
+
+			let code_len = module.code.len();
+			// We print this to measure the impact of the migration.
+			// Storage removed: deleted PrefabWasmModule's encoded len.
+			// Storage added: determinism field encoded len (as all other CodeInfo fields are the
+			// same as in the deleted OwnerInfo).
+			log::debug!(target: LOG_TARGET, "Storage removed: 1 item, {} bytes", &code_len,);
+
+			// Storage usage prices could change over time, and accounts who uploaded their
+			// conctracts code before the storage deposits where introduced, had not been ever
+			// charged with any deposit for that (see migration v6).
+			//
+			// This is why deposit to be refunded here is calculated as follows:
+			//
+			// 1. Calculate the deposit amount for storage before the migration, given current
+			// prices.
+			// 2. Given current reserved deposit amount, calculate the correction factor.
+			// 3. Calculate the deposit amount for storage after the migration, given current
+			// prices.
+			// 4. Calculate real deposit amount to be reserved after the migration.
+			let price_per_byte = T::DepositPerByte::get();
+			let price_per_item = T::DepositPerItem::get();
```

Comment:
Why don't make it independent from the current prices by doing it pro rata? This is how we are doing it for the normal storage deposits. What I mean: Calculate how much % of storage was removed. Then just remove that ratio of reserved balance.

Reply:
Hm basically IIUC it should give the same result. But I agree that the way we do it in the storage meter is simpler. And it's better to put it clear we use the same approach to pro rata storage deposits refunding everywhere codewise. Will change this.

Reply:
@athei Aghh no we can't do that, because the deposit is two-dimensional, hence for that we would need to know which part of the deposit amount is for items and which is for bytes, but this information isn't available at this point. That's why we have to calculate it this way here.

Reply:
@athei Aghh no we can't do that, because the deposit is two-dimensional, hence for that we would need to know which part of the deposit amount is for items and which is for bytes, but this information isn't available at this point. That's why we have to calculate it this way here.

Reply:
The complexity of the calculation and that it requires the byte prices not to change makes me a bit uneasy. I would say just refunding 50% and calling it a day is good enough. It is okay because the real refund would be  even more.

Reply:
> I would say just refunding 50% and calling it a day is good enough. It is okay because the real refund would be even more.

No, real refund is less than 50%, even theoretically. This is again because we have the part charged for the storage items, as
`(1/2 +p)x + 2/3 y > 1/2 (x + y)` is true for every non-zero `y`. Here we designate `x` and `y` are per byte and per item prices, and `p` is overhead gas metering instrumentation adds to the module size.

This is also confirmed by the data we get from the try-runtime logs (see PR desc) where it results in ~40% total deposit refund.

Reply:
> I would say just refunding 50% and calling it a day is good enough. It is okay because the real refund would be even more.

For the real refund to be more than 50%, theoretically this statement should be true 

`(1/2 + p)x + 1/3 y > 1/2 (x + y)`

which gives us 

`px > 1/6 y`

Here we designate `x` and `y` are per byte and per item prices, and `p` is overhead gas metering instrumentation adds to the module size.

So 

`6p > y/x`

Currently `y/x` [is](https://paritytech.github.io/substrate/master/src/kitchensink_runtime/constants.rs.html#29) `15/6 = 2.5` therefore the refund would be more than half if `6p > 2.5 => p > 0.41`.

While this can be true for some contracts, it is not the case for the median contract deployed on Rococo, which confirmed by the data we get from the try-runtime logs (see PR desc) where it results in ~42% total deposit refund.

Reply:
For the reference, 
<details>
<summary>there is the size overheads made by the instrumentation for different fixtures</summary>

```
| fixture                      |  original size   | gas metered/host fn |
|------------------------------|------------------|---------------------|
| recursive_ok.wat             |             0 kb |         0 kb (137%) |
| count_until.wat              |             0 kb |         0 kb (125%) |
| global_bump.wat              |             0 kb |         0 kb (123%) |
| memory-vec-add.wat           |             0 kb |         0 kb (116%) |
| factorial.wat                |             0 kb |         0 kb (125%) |
| fibonacci.wat                |             0 kb |         0 kb (121%) |
| contract_terminate.wasm      |             1 kb |         1 kb (110%) |
| coremark_minimal.wasm        |             7 kb |         8 kb (114%) |
| trait_erc20.wasm             |            10 kb |        11 kb (108%) |
| rand_extension.wasm          |             4 kb |         5 kb (109%) |
| multisig.wasm                |            27 kb |        30 kb (110%) |
| wasm_kernel.wasm             |           779 kb |       787 kb (100%) |
| many_blocks.wasm             |          1023 kb |      2389 kb (233%) |
| contract_transfer.wasm       |             7 kb |         8 kb (113%) |
| erc1155.wasm                 |            26 kb |        29 kb (111%) |
| erc20.wasm                   |             9 kb |        10 kb (108%) |
| dns.wasm                     |            10 kb |        11 kb (108%) |
| proxy.wasm                   |             3 kb |         4 kb (108%) |
| erc721.wasm                  |            13 kb |        14 kb (108%) |
```
</details>

where we see that `p > 0.41` holds only for the single fixture `many_blocks.wasm`